search/api.py

```diff
@@ -1,306 +1,142 @@
-            'start_date': DATE_TYPE,
-            'state_or_territory': NOT_ANALYZED_STRING_TYPE,
-        }},
-    })
-    mapping.field("program", "object", properties={
-        'id': LONG_TYPE,
-        'grade_average': LONG_TYPE,
-        'enrollments': {'type': 'nested', 'properties': {
-            'course_details': {
-                'type': 'object',
-                'properties': {
-                    'course_modes': {
-                        'type': 'nested',
-                    }
-                }
-            }
-        }},
-        'certificates': {'type': 'nested'}
-    })
-    # Make strings not_analyzed by default
-    mapping.meta('dynamic_templates', [{
-        "notanalyzed": {
-            "match": "*",
-            "match_mapping_type": "string",
-            "mapping": NOT_ANALYZED_STRING_TYPE
-        }
-    }])
-    return mapping
+    def add_to_results(self, es_search_hits):
+        """
+        Adds a set of search results ('hits') to the combined results
+        """
+        self.results = self.results + list(es_search_hits)
 
+    def execute(self, search_obj, page_size=DEFAULT_ES_LOOP_PAGE_SIZE):
+        """
+        Executes a search query, pages through all of its results, maintains
+        a record of all results found, and returns those combined results.
+        """
+        self.init_combined_results()
+        # Maintaining a consistent sort on '_doc' will help prevent bugs where the
+        # index is altered during the loop
+        search_obj.sort('_doc')
+        loop = 0
+        all_results_returned = False
+        while not all_results_returned:
+            from_index = loop * page_size
+            to_index = from_index + page_size
+            search_results = execute_search(search_obj[from_index: to_index])
+            self.add_to_results(search_results)
+            all_results_returned = to_index >= search_results.hits.total
+            loop += 1
+        return self.results
 
-def create_program_enrolled_user_mapping():
-    """
-    Save the mapping for a program user. If one already exists, delete it first.
-    """
-    conn = get_conn(verify=False)
-    index_name = settings.ELASTICSEARCH_INDEX
-    if conn.indices.exists_type(index=index_name, doc_type=USER_DOC_TYPE):
-        conn.indices.delete_mapping(index=index_name, doc_type=USER_DOC_TYPE)
-    mapping = program_enrolled_user_mapping()
-    mapping.save(index_name)
+    # Make the objects callable - ie: invoking an object of this class directly
+    # will call the given method.
+    __call__ = execute
 
 
-def create_mappings():
+class ESFullResultEmailSearch(ESFullResultSearch):
     """
-    Create all mappings, deleting existing mappings if they exist.
+    Provides a function to return all unique emails for documents that match
+    an ES query
     """
-    create_program_enrolled_user_mapping()
+    def init_combined_results(self):
+        self.results = set()
 
+    def add_to_results(self, es_search_hits):
```

Comment:
sorry, didn't see the comments in response. `get_all_query_matching_emails` _is_ effectively a function. you can call it, and indeed it is being called in mail/views.py. what we get from the inheritance relationship is code reuse and, in general, the representation of a concept in code. whether we want to retrieve (a) the full ES documents that match a query, or (b) the full set of email addresses in those documents that match a query, that process is very similar. in both cases (to quote the docstring), we "execute a search query, page through all of its results, maintain a record of results found, and return those combined results". what's different between a and b? (1) the data structure that is most appropriate to store the results, and (2) the way that results are added when a page of ES documents is retrieved. hence, the two methods that are implemented differently in those two classes. in the case of `get_all_query_matching_emails`, duplicates could indeed be removed as part of `execute`, but that duplicate removal is irrelevant to `get_all_query_matching_results`. `execute` is largely the same process for both, so inheritance/code reuse is appropriate

if you still don't agree with or understand what i'm going for, i'm happy to chat in person